### PR TITLE
fix: Issue in non-Tensor Input Resolution

### LINKED ISSUE #1612: 🐛 [Bug] Node Inputs Cannot be Evaluated at Conversion Time
See NVBugs

### PR DESCRIPTION
**Pull request by gs-olive**
# Description

- <s>In certain TensorRT-executed blocks, multiple non-Tensor inputs originate from the same parent Tensor in another Torch block, so a single pass of resolveTRTNonTensorInputs does not resolve all non-Tensor inputs, causing failures at compile-time</s>
- Issue was determined to trace to the use of `prim::Loop` in the TensorRT block and is fixed in PR #1691 
- Add function to `SegmentedBlock` to get the ID of a block, so re-inserted blocks retain the same ID as the previous block being replaced
- Add regression test case to elicit behavior

### Before Non-Tensor Input Resolution
```python
DEBUG: [Torch-TensorRT - Debug Build] - Finalizing in progress TensorRT block
DEBUG: [Torch-TensorRT - Debug Build] - Segment Block @2:
    Target: TensorRT

    Graph: graph(%1 : int,
      %9 : int[],
      %12 : Tensor):
```
### After 1 Round of Non-Tensor Input Resolution
```python
GRAPH: [Torch-TensorRT - Debug Build] - Running shape analysis on block Segment Block @2:
    Target: TensorRT

    Graph: graph(%1 : int[],
      %3 : Tensor):
```

### After 2 Rounds of Non-Tensor Input Resolution
```python
INFO: [Torch-TensorRT - Debug Build] - Segment Block @2:
    Target: TensorRT

    Graph: graph(%1 : Tensor):
```

Fixes #1612 
Related to: #1613 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [ x ] My code follows the style guidelines of this project (You can use the linters)
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ x ] I have made corresponding changes to the documentation
- [ x ] I have added tests to verify my fix or my feature
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

**Comment by peri044** [comment]
Expected this PR to make it into 23.03 release

**Comment by bowang007** [comment]
This does not make sense in fact. 
Firstly, even if 2 non-Tensor inputs come from a same node, that node will be added into the segmented block. There is no need to do it for multiple passes. 
Secondly, I guess why it fails for a single pass is because that there is a Loop block in the TensorRT segment. This could introduce undefined behaviors. The real issue I think here is the Loop block, it shouldn't be in TensorRT segment at all. In our last code refactoring for partitioning, we set all nodes aggressively to be run in TensorRT, this introduces some issue like LOOP nodes default to be run in TensorRT, this could result in some issues like this.

File: core/partitioning/partitioning.cpp [hunk 1]
```diff
@@ -286,7 +286,8 @@ void resolveTRTNonTensorInputs(PartitioningCtx* ctx, torch::jit::Block* block) {
             dependency_nodes.end(),
             cur_partitioned_block[i].raw_nodes().begin(),
             cur_partitioned_block[i].raw_nodes().end());
-        cur_partitioned_block[i] = SegmentedBlock(SegmentedBlock::kTensorRT, dependency_nodes);
+        cur_partitioned_block[i] =
+            SegmentedBlock(cur_partitioned_block[i].get_id(), SegmentedBlock::kTensorRT, dependency_nodes);
```
Comment (gs-olive) [inline core/partitioning/partitioning.cpp:290]: Added to ensure the block IDs of segmented blocks stay in order despite resolution of non-Tensor inputs.

**Review by bowang007 — APPROVED** [review]
LGTM, please squash your commits then merge.

**Pull request merged**